### DISCUSSION TITLE
Версия библиотеки 2.0 (И Яндекс Карты 3.0)

### DISCUSSION THREAD

**Discussion by daniluk4000** [yandex-maps-unofficial/vue-yandex-maps#60] [Анонсы]
Всем привет! Для начала, хотел бы представиться - меня зовут Данил и я буду помогать @PNKBizz с обслуживанием этой библиотеки, я же сделал версию либы 2.0 (и буду отвечать на вопросы/баги по ней).

## Текущий статус
Выпущен релиз кандидат.

[Документация](https://yandex-maps-unofficial.github.io/vue-yandex-maps/).

Статус готовности к релизу можно отслеживать в [Milestone 2.0](https://github.com/yandex-maps-unofficial/vue-yandex-maps/milestone/2).

## Что такое vue-yandex-maps 2.0
Библиотека была, по сути, написана с нуля. 

## Что изменилось
- Переезд на Яндекс Карты 3.0
- Поддержка официальных @yandex/ymaps3-types и обновлений независимо от нашей библиотеки (кроме новых компонентов)
- Поддержка Vue2/Vue3 в рамках одной версии
- Полная поддержка Nuxt 2/Bridge/3, включая конфигурацию из Nuxt Config
- Поддержка SSR
- Полная поддержка TypeScript, в том числе при задании настроек
- Переход на ES Modules
- Отказ от Event Emmitter в пользу реактивных ref в namespace проекта
- Поддержка разных способов загрузки карты
- Загрузка карты автоматически при первом вызове компонента Карт

### Изменения по сравнению с 1.0 (и другими)
- Полная реактивность настроек компонентов
- Поддержаны все компоненты, документированные Яндексом в Картах 3.0
- Слот Маркера теперь реактивен, все изменения внутри него обновляются в Картах мгновенно

### Улучшения по сравнению с Картами 3.0 (vanilla)
- Вы можете использовать многие классы оттуда как компоненты (как в их библиотеке для React, за парой отличий)
- При использовании компонента, для доступа к которому необходимо сделать импорт из Карт, импорт произойдет автоматически (например, так работают ползунки зума и пр.)

Изменения в Картах 3.0 [читайте на сайте Яндекса](https://yandex.ru/blog/mapsapi/vyshel-javascript-api-3-0-bolee-bystraya-versiya-programmnoy-biblioteki-na-baze-vektornykh-kart).

## Процесс релиза

В рамках релиза я планирую идти по схеме, где:
- Версия alpha не содержит документации и полного набора компонентов. Это внутренняя версия для тестирования на различных проектах мной
- Версия beta идёт с документацией, но без Migration Guide и, вероятно, примеров работы. Она содержит все компоненты и готова к тестированию
- Версия rc (Release Candidate) является полностью готовым и документированным продуктом
- Релизная версия отличается от RC исключительно исправлением ошибок

Релизную версию я планирую выпустить до конца года.

### Почему так долго?
Библиотеку я пишу в свободное время по выходным.

### Тяжело будет переходить?
Да. Готовьтесь к тому, что код нужно будет писать с нуля. Я постараюсь сделать настолько подробный Migration Guide, насколько это возможно, но всего он не покроет. 

Причина этому не столько то, что всё было написано с нуля, сколько полностью изменившееся API Яндекса. Можете почитать их [Migration Guide](https://yandex.ru/dev/jsapi30/doc/ru/upgrade/) - он, мягко говоря, не покрывает всего, что они поменяли.

### Что будет с репозиторием vue-yandex-map и версией 1.0.0?
Решение примет @PNKBizz, но в моём понимании 1.0.0 прекратит поддержку, а репозиторий vue-yandex-map станет readonly и будет вести на этот репозиторий через Readme.

### Когда будет бета?
Бета-версию я планирую выпустить в сентябре-октябре.

### Если тебя собьет машина, то как библиотека будет поддерживать актуальность?
Большинство настроек сделано таким образом, что вы сможете самостоятельно обновлять версию @yandex/ymaps3-types и получать актуальную типизацию. В том числе поэтому настройки передаются через settings, а не отдельные props.

В случае, если Яндекс добавит новые компоненты, и некому по каким-то причинам будет добавить их в библиотеку (по ряду причин), вы всегда сможете добавить компоненты через инстанс карты (addChild/addLayer и пр.), как указано в документации Яндекса.

### Vue 2 будет поддерживаться вечно?
Нет. Данная версия поддерживает Vue 2/Nuxt 2/Nuxt Bridge через жертву в виде написания кода template через return в setup. Любые другие решения не работают и теряют совместимость.

В таком виде поддержку планирую продолжать как максимум до момента, пока библиотеку будет невозможно поддерживать и сохранять обе версии. После этого версию для Vue 2 придется дропнуть, а Vue 3 переписать на нормальную интеграцию с `<template>` и потенциально `<script setup>` в версии по типу v3.

Также рассмотреть дроп поддержки планирую после официального окончания поддержки Vue 2 (декабрь 2023) + еще несколько месяцев.

После дропа поддержки будет выпущена версия 3.0, где будут внесены Breaking Changes (по необходимости) и внедрены нормальные template.

Цель же сейчас - дать пользователям Vue 2 возможность использовать Карты 3.0 (а заодно избавиться от страданий в виде двух разных версий пакетов).

----

Данный пост я написал, чтобы немного посвятить вас в свои планы, так как, уверен, Карт 3.0 ждут многие. Лично мне они понравились своей скоростью работы и возможностью нормально сделать реактивность. Код на них, пока что, писать было удовольствием.

Если у кого-то здесь есть ещё вопросы, буду рад на них ответить :)

**Comment by simensBat** [comment]
@daniluk4000 Привет. Хотел поиграться с текущим кодом, но у меня не получилось запустить карты на nuxt 2 и vue 2. 

Первая ошибка была в функции `isDev` в `utils.ts`

```typescript
export function isDev() {
  // @ts-expect-error
  return (typeof process !== 'undefined' && process.env?.NODE_ENV === 'development') || (import.meta?.env?.DEV);
}
```

Похоже `import.meta?.env?.DEV` для 2 nuxt не работает, из за этого билд не собирается, как я понял это связано с webpack 4, он не умеет работать с import meta. Если эту проверку убрать, билд собирается.

Дальше я хотел посмотреть на работу карты, но у меня падает страница с ошибками:

1) "The "data" option should be a function that returns a per-instance value in component definitions." "YandexMap"
2) Failed to mount component: template or render function not defined.

Как я понял это связано с тем, что вы писали выше "Данная версия поддерживает Vue 2/Nuxt 2/Nuxt Bridge через жертву в виде написания кода template через return в setup. Любые другие решения не работают и теряют совместимость" 

Пробую возвращать в функции setup() рендер функцию, но она не исправляет ситуацию, ошибки такие же. Пример кода.

```typescript
import { defineComponent, h } from 'vue';
import {
  YandexMap,
  // YandexMapDefaultFeaturesLayer
  YandexMapDefaultSchemeLayer,
} from 'vue-yandex-maps';

export default defineComponent({
  setup() {
    return () => {
      // @ts-expect-error
      return h(YandexMap, {
        props: {
          width: '50dvw',
          height: '75dvh',
          settings: { location: { center: [37.588144, 55.733842], zoom: 7 } } 
        }
      // @ts-expect-error
      }, [h(YandexMapDefaultSchemeLayer, {}, {props: {settings: { theme: 'dark' }}})])
    } 
  }
});
```
Как можно исправить проблему, чтоб посмотреть на работу карты?

**Reply by daniluk4000** [reply]
Привет-привет! 

Подскажите пожалуйста, у вас Vue 2.7? Ошибки очень странные.

Если ниже 2.7, то установлен ли модуль для composition API? Без него на 2.6 и ниже не зафурычит.

Еще просьба подсказать, какая была ошибка в isDev. 

По поводу поиграться: советую еще глянуть packages/examples, если не смотрели. В examples/vue3 есть example-component с базовыми примерами кода, в vue2/nuxt2 можно посмотреть, как инициализировать и работать. 

Не забудьте для токена в кабинете Яндекса указать домены. Без этого также не заведется. Это всё покрою в документации, когда буду оформлять beta-сборку =)

**Reply by simensBat** [reply]
Версия Vue 2.7.14. Модуль установлен.  10 Винда, 20 нода, chrome. Что я делал: 
1) Cклонировал вашу ветку v2
2) Установил зависимости yarn под 20 ноду `yarn`

![image](https://github.com/PNKBizz/vue-yandex-maps/assets/61266300/2a44e232-4f32-4398-a715-4be8bff6a437)

3) Сбилдил пакет либы `yarn build`
4) Пробую запускать nuxt 2 для теста `yarn workspace example-nuxt2 dev`
5) Проект собрался,  есть ошибки в консоле. Это если поправить код `isDev`

![image](https://github.com/PNKBizz/vue-yandex-maps/assets/61266300/dee8bde8-2e6a-4c5d-90d7-99fb19412823)


![image](https://github.com/PNKBizz/vue-yandex-maps/assets/61266300/39c7b099-5757-44b1-b810-4dad31d1dde5)

Ошибка при сборке, если не править isDev

![image](https://github.com/PNKBizz/vue-yandex-maps/assets/61266300/cf3cbb1e-6bab-456c-bdeb-f2c8d5bc1b52)

Для своего примера я убрал почти все, чтоб попытаться понять в чем проблема, взял только YandexMap и слой, чтоб запустить карту.

Про токен вкурсе, с 3 версией  карт уже работал. Решил пока свой не ставить, т.к если запустить 3 пример карт на vue 3, они работают и там карта грузится. Стоит ваш токен, значит сделал вывод, что дело не в нем.

**Reply by daniluk4000** [reply]
Очень интересно. В выходные постараюсь посмотреть. Возможно, я что-то сломал, оно работало))

Спасибо!

В целом я карты уже заводил на реальном Nuxt 3 проекте, на Vue 2 честно пока только в демке, но существенных различий быть не должно.

**Reply by daniluk4000** [reply]
Нашел проблему. Vue 2 не любит такие манипуляции со слотами, которые делал я :)
В целом я его понимаю. Мне бы тоже не нравилось.

Также почитал, Evan You говорит, что общение между слотом и родителем через эмит слишком непредсказуемое, поэтому они это не реализовали. А я судя по всему сделал что-то не совсем законное :D

Остальные ошибки также починил. Опубликовал 2.0.0-alpha.15.

**Reply by simensBat** [reply]
Да сейчас на nuxt 2 запускается. Хотел еще один вопрос прояснить. 

У Vue 2.7 есть [нативная поддержка](https://v2.vuejs.org/v2/guide/migration-vue-2-7#Behavior-Differences-from-Vue-3) composition api, исходя из этого возникает вопрос, можно ли убрать плагин `@vue/composition-api` для версий nuxt  >= 2.6.13(вроде там добавили Vue 2.7 в ядро nuxt) или для vue >= 2.7?

Для теста решил проверить пример `example-nuxt2`, убрал из него все лишние зависимости.

![image](https://github.com/PNKBizz/vue-yandex-maps/assets/61266300/e32da5ee-5858-4867-9d0c-627cdd470f2b)

![image](https://github.com/PNKBizz/vue-yandex-maps/assets/61266300/acbc8ecd-cd6a-474d-a7f5-cdf76b3bc78d)

Как итог карта не запустилась.

![image](https://github.com/PNKBizz/vue-yandex-maps/assets/61266300/f0dbcccd-aaa7-4580-a29d-41a6b7de422b)

Предупреждения появляются внутри `setup` функции в компоненте `YandexMap.vue`. Странные ошибки честно говоря.

![image](https://github.com/PNKBizz/vue-yandex-maps/assets/61266300/257276ad-d4e3-4b67-a607-06d229e9d482)

![image](https://github.com/PNKBizz/vue-yandex-maps/assets/61266300/ac51edc8-81c4-4f96-b31b-5ce1b4d2c1a9)

**Reply by daniluk4000** [reply]
Тут всё просто.
1. Да, для Vue 2.7 плагин @vue/composition-api не нужен
2. Но вот для Nuxt 2 @nuxtjs/composition-api нужен обязательно, кроме версии Bridge, потому что по другому Nuxt 2 не заведется даже на Vue 2.7

**Comment by vermorag** [comment]
@daniluk4000 Добрый день. Тоже решил провести эксперимент с картой на vue3 - есть потребность поменять location у нее после монтирования (location прокинут через pinia хранилище). В коде вижу отсылку, что settings не реактивно и надо пользоваться setLocation. Однако с кодом ниже ничего не работает:
```
<script lang="ts" setup>
    import { YandexMap, YandexMapDefaultSchemeLayer, YandexMapDefaultFeaturesLayer } from 'vue-yandex-maps';
    import useStorage from '../store/map';
    import { toRaw, watch } from 'vue';
    import { YMap } from 'ymaps3/*';

    let map: YMap | null;

    const mapStorage = useStorage();

    watch(mapStorage.location, () => {
        if (map) {
            map.setLocation(toRaw(mapStorage.location));
        }
    });
</script>

<template>
    <YandexMap :settings="{ location: mapStorage.location }" @update:modelValue="(e: YMap) => (map = e)">
        <YandexMapDefaultSchemeLayer />
        <YandexMapDefaultFeaturesLayer />
    </YandexMap>
</template>
```
watch отрабатывает, метод setLocation вызывается, но карта остается там же, где и была. Других методов как мне получить рабочий экземпляр карты и у него поменять параметры я не нашел. Можешь подсказать?

**Reply by daniluk4000** [reply]
Привет! 

1. map точно изменяется, да? Например, на моём личном проекте map является `const map = shallowRef<null | YMap>(null);`
2. В консоли ошибок нет? 
3. mapStorage.location валиден для Яндекса?

Почему так настойчиво спрашиваю - код рабочий. 

![image](https://github.com/PNKBizz/vue-yandex-maps/assets/10450717/3e40af89-cb05-4f3a-9661-e17cd3051594)

**Reply by daniluk4000** [reply]
Настройки у (почти) всего остального будут реактивны, это особенность именно инстанса карты от самого Яндекса.

**Reply by vermorag** [reply]
Ну, добавил еще console.log вот так:
```
    watch(mapStorage.location, () => {
        if (map) {
            console.log(map);
            console.log(toRaw(mapStorage.location));
            map.setLocation(toRaw(mapStorage.location));
        }
    });
```
И он мне великолепно дампит объект карты и локейшен.
![image](https://github.com/PNKBizz/vue-yandex-maps/assets/25276845/d5b81253-26a7-4c0c-8fae-a1c59699370a)

**Reply by vermorag** [reply]
Переделал на shallowRef - стало отрабатывать, но только самое первое изменение, пришедшее из хранилища. А все остальные изменения игнорирует. Вообще не понимаю как так.

**Reply by daniluk4000** [reply]
Оочень странные дела. Тем более, что у меня примерно аналогичный код работает многократно.

Сможете портировать свой код в какой нибудь репозиторий? Мне нужен reproduction)

Я сам попытаюсь воспроизвести, но не пойму, где оно ломается у вас...

Еще могу предложить переделать код на чистый YMap и попробовать поработать с ним напрямую (имею в виду без библиотеки, чисто по доке Яндекса). Если там ошибки не будет - уже грешить на либу

Карты еще очень капризные. Что делает toRaw, к слову, не подскажете?

И попробуйте для эксперимента еще setLocation({ coordinates: mapStorage.location.coordinates }) - будет ли меняться, если не весь объект ему передавать

**Reply by daniluk4000** [reply]
@vermorag разобрались?

**Reply by simensBat** [reply]
Была такая же проблема с setLocation(), если передавать не объект, а координаты, перехода не происходит. Нужно передавать: 

```typescript
mapInstance.value?.setLocation({ center: [0, 0] })
```

**Reply by simensBat** [reply]
@daniluk4000 Заметил еще такую особенность на Vue 2.7. Если открыть девтулзы vue на карте у которой больше 2000 маркеров с кластерами, карта начинает жутко лагать и память на вкладку увеличивается постепенно. Как я понимаю это связано с тем, что в девтулзах создается огромное дерево компонентов на каждый маркер и кластер, из за этого лаги происходят.

Такая же проблема была и на первой версии библиотеки. Там это отчасти решалось опцией objectManager,  который помогает рисовать карту, если на ней много маркеров.

Хоть это не проблема библиотеки получается но все-равно не приятно дебажить код.

**Reply by daniluk4000** [reply]
@Batyodie ну вы 2 тысячи не рендерите пожалуйста и всё будет хорошо

Шутка) Это только с девтулами? Если да, то жить можно. Вероятно, это особенность Карт.

Еще дело в том, что компонент кластера в Картах 3.0 написан на лютейших костылях, потому что сами Яндекс не смогли его нормально сделать. Вроде бы кластера даже в примерах у них нет.

Я решил, что компонент жизненно важен (еще бы), поэтому портировал его. Я мог где-то ошибиться в реализации, поэтому главное, чтобы не было утечек памяти БЕЗ девтулов и в продакшн-режиме.

**Reply by vermorag** [reply]
@daniluk4000 Не разобрался. Объект mapStorage.location работает с яндексовской оберткой под реакт. toRaw - снимает прокси с объекта, чтобы яндекс его мог спокойно переварить (у реакта была эта проблема, когда ему скармливаешь прокси, поэтому предпочитаю кастовать). Но переделал вызов также, как его сделал в реакте - заработало.
```
    watch(mapStorage.location, val => {
        map?.value?.setLocation({...val});
    });
```
А почему мы не можем сделать props с локейшеном (как в реактовской обертке) и в компоненте также [отслеживать](https://github.com/vermorag/vue-yandex-maps/blob/v2/packages/vue-yandex-maps/src/components/YandexMap.vue#L133) его изменение и вызывать setLocation? Ну просто потому что это достаточно часто используемая операция с картой и каждый раз отлавливать карту по евенту апдейта и потом обновлять - не сильно весело.

Кластер в примерах у яндекса есть, но он немного странный в использовании. https://yandex.ru/dev/jsapi30/doc/ru/ref/packages/clusterer/

**Reply by daniluk4000** [reply]
Это не совсем пример (не находится в основной секции), но да, отсюда и черпал.

> А почему мы не можем сделать props с локейшеном

В setLocation еще есть прекрасная анимация. Мне не сложно сделать пропс, но как будто setLocation более выгодно использовать.

Карту по ивенту отслеживать не надо, есть v-model.

> Но переделал вызов также, как его сделал в реакте - заработало

То есть вместо toRaw сделали спред-оператор? В общем я не знаю, у меня всё работало и без этого. Тут я не могу повлиять особо, метод передается напрямую в Яндекс. Равзе что вотчер сделать на настройки, да.

**Reply by daniluk4000** [reply]
Я просто не особо хотел бы "расслаблять" пользователей библиотеки тем, что локейшн меняется сам по себе, потому что ну с анимацией намного лучше это выглядит)

**Reply by daniluk4000** [reply]
Ну и по такой логике надо весь код будет покрыть setBehaviours и т.д.

Если будет запрос от сообщества, то сделать реально.

**Reply by vermorag** [reply]
setLocation - он и анимацию же поддерживает, если ему передавать значение анимации. Тут не понимаю в чем вопрос.

**Reply by daniluk4000** [reply]
Непонятно, как сделать передачу анимации (длительность и тип), если location будет отслеживаться по settings.location

**Reply by vermorag** [reply]
Изменение settings.location  вызывает отработку метода setLocation.
Т.е. мне достаточно изменить settings.location на 
```
{
  "center": [10,10],
  "duration": 1000,
  "easing": "linear"
}
```
И карта поменяется с анимацией.
Собственно, у меня изначально location в примерах и содержит значение длительности, чтобы все было плавно. И это работает.
![275705357-d5b81253-26a7-4c0c-8fae-a1c59699370a](https://github.com/PNKBizz/vue-yandex-maps/assets/25276845/446d1970-d8eb-466b-9882-34343e17270e)

**Reply by daniluk4000** [reply]
Таким образом я отбираю свободу у разработчика.

В общем я запишу себе в туду, это вполне реально поработать так, чтобы было кайфом.

То, что оно работает, я знаю)

**Reply by vermorag** [reply]
И чем отбирается свобода? 
Передавать или нет доп параметры - это же выбор разработчика.

**Reply by daniluk4000** [reply]
А, имеется в виду передача настроек анимации в location прям в settings

Тут есть другой момент - нужно будет модифицировать немного объект настроек: по хорошему не передавать поля с анимацией при инициализации карты, кроме того, надо будет расширить YMapProps и заменить там location на объект настроек setLocation - иначе тайпскрипт не будет рад

Как и сказал, запишу в туду)

**Reply by daniluk4000** [reply]
@vermorag https://github.com/yandex-maps-unofficial/vue-yandex-maps/issues/75

До RC сделаю

**Reply by daniluk4000** [reply]
Смешная история, на самом деле. Делал вчера новые примеры и решил заодно это сделать. 

Вы упоминали, что Яндекс в своей обертке для реакта сделали это, но на самом деле в примере они всё меняют вручную через setLocation/setBounds и тд) Это забавно

И да - там Яндексом всё из коробки поддерживается, всё лучше, чем я думал, даже типы не пришлось менять. Немного подхимичил с изменением location/camera, в остальном реализация полностью нативная.

[Вот тут документировал](https://yandex-maps-unofficial.github.io/vue-yandex-maps/components/map.html#settings)

**Comment by simensBat** [comment]
@daniluk4000 Есть не только у дев тулзов проблема с производительностью. Возможно это только на vue 2.7 или я где-то в коде косяк сделал, но плавность при раскрытии большого кластера уходит. Я открыл условно Питер, где очень много маркеров и кластеров, выбрал большой кластер, раскрыл его(сделал зум в область кластера, часть маркеров из кластера показались) и плавность ушла. Вместо 60 кадров стало 30.

Далее эти 30 кадров уходили на другие города, если я просто меняю локацию и количество кластеров и маркеров. 

Чтобы починить плавность для других локаций, где маркеров может быть мало,  нужно убивать карту и заново ее рендерить.

**Reply by daniluk4000** [reply]
Яндекс очень много кэширует в этом плане. Есть шанс, что ваши маркеры были закэшированы, а так как их много, это повлияло на перфоманс.

Скоро мы дадим возможность заводить ишью на v2. Предлагаю вам завести ишью на оптимизацию кластера, в идеале с reproduction-репой или кодом. А я подумаю, что можно сделать. Я там уже нарушаю правила и модифицирую кэш Яндекса (без этого я не смог его завести с реактивностью в принципе), может и перфоманс смогу поднять.

**Reply by daniluk4000** [reply]
@Batyodie 

1. Нашел проблему с перфомансом. Это из-за того, как именно библиотека рисует компоненты. По сути, мы позволяем вам максимально удобно писать код, но из-за этого все компоненты кладутся прямо в DOM. Посмотрел, как сделано у Яндекса, в целом можно повеселиться и сделать так же. Завел ишью на разработку режима с высокой производительностью: https://github.com/yandex-maps-unofficial/vue-yandex-maps/issues/81
2. По основному багу, по ощущениям, всё же связано с кэшированием от Яндекса. Если у вас проблема сохраняется с новым апдейтом, заведите, пожалуйста, ишью, я попытаюсь выкинуть кэш Яндекса целиком

**Reply by daniluk4000** [reply]
@Batyodie и это снова я. Прокачал кластер. 5 тысяч элементов щелкает как орешки. Выкатил в `2.0.0-beta.7`.

**Reply by simensBat** [reply]
@daniluk4000 Привет. Круто! Как будет время, посмотрю. Есть сейчас понимание, когда проект можно будет поставить через npm? Сейчас не очень удобно его тестировать.

**Reply by daniluk4000** [reply]
@Batyodie а вы как ставили?) Сочувствую

Версия уже в npm под тэгом `next`

https://yandex-maps-unofficial.github.io/vue-yandex-maps/guide/vue3.html

**Reply by simensBat** [reply]
@daniluk4000 Понял, не знал. Я билдил проект локально в ручную, потом исходники переносил куда мне нужно.

**Comment by aleksei-deets** [comment]
Добый день. 
Подскажите пожалуйста, есть ли у бибилиотеки поддержка Vue фреймворка Quasar Framework?
